src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -67,190 +68,186 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
-
-                    // We do not differentiate array of different kinds sicne they are all represented in the indices as "NonArrayElementTypeName[]"
-                    // e.g. int[], int[][], int[,], etc. are all represented as "int[]", whereas array of complex type such as T[] is "[]".
-                    return elementTypeName + FindSymbols.Extensions.ArrayReceiverTypeNameSuffix;
-
-                default:
-                    // Complex types are represented by "";
-                    return FindSymbols.Extensions.ComplexReceiverTypeName;
+                    overloadMap[(containingNamespacename, symbol.Name, symbol.Arity > 0)] = (bestSymbol, overloadCount);
+                }
             }
-        }
 
-        public static async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)> GetUnimportedExtensionMethodsInCurrentProcessAsync(
-            Document document,
-            int position,
-            ITypeSymbol receiverTypeSymbol,
-            ISet<string> namespaceInScope,
-            bool forceIndexCreation,
-            CancellationToken cancellationToken)
-        {
-            var counter = new StatisticCounter();
-            var ticks = Environment.TickCount;
+            // Then convert symbols into completion items
+            using var _3 = ArrayBuilder<SerializableImportCompletionItem>.GetInstance(out var itemsBuilder);
 
-            var indicesResult = await TryGetIndicesAsync(
-                document.Project, forceIndexCreation, cancellationToken).ConfigureAwait(false);
+            foreach (var (methodData, overloadSymbols) in overloadMap)
+            {
+                // To display the count of of additional overloads, we need to substract total by 1.
+                var item = CreateItem(overloadSymbols.bestSymbol, methodData.containingNamespace, additionalOverloadCount: overloadSymbols.overloadCount - 1, cancellationToken);
+                itemsBuilder.Add(item);
+            }
 
-            // Don't show unimported extension methods if the index isn't ready.
-            if (!indicesResult.HasResult)
+            // If we don't have all the indices available already, queue a backgrounds task to create them.
+            if (!indicesComplete)
             {
-                // We use a very simple approach to build the cache in the background:
-                // queue a new task only if the previous task is completed, regardless of what
-                // that task is.
                 lock (s_gate)
                 {
+                    // We use a very simple approach to build the cache in the background:
+                    // queue a new task only if the previous task is completed. This is to avoid
+                    // queueing calculation for the same set of references repeatedly while
+                    // index is being constrcuted, which might take some time.
                     if (s_indexingTask.IsCompleted)
```

Comment:
i don't get the relation between teh `indicesComplete` flag and the s_indexingTask.

Reply:
`indicesComplete` is `true` means we have indices available for all references. When it's `false`, we would try to queue a background task to force create them, which is `s_indexingTask`, however, if `s_indexingTask` from a previous completion session is not completed, we will just let it finish w/o attaching another task to it.